### PR TITLE
Add script to run a local network for HF testing

### PR DESCRIPTION
**Pull request by georgeee**
1. Update prepare-test-ledger.sh to allow generating a ledger with normal distribution, delegatee balance cutoff and a nicer CLi interface
2. Introduce a new script that generates a mainnet-like ledger with all stake delegated to a key and start a network of two nodes operating off this ledger

To learn about usage of both scripts, run them with `-h` flag.

Script `scripts/run-hf-localnet.sh` can be used without change with both `compatible` and `berkeley`. In latter case `-b` flag should be provided.

Script `scripts/run-hf-localnet.sh` generates a configuration based on mainnet ledger and launches two local nodes.
It also sends transactions on interval to simulate networking activity and ledger mutation.

Usage as simple as (launches a local node with the given mina executable, if not provided `mina` command will be used):
```
./scripts/run-hf-localnet.sh -m _build/default/src/app/cli/src/mina.exe -b
```

Explain how you tested your changes:
* Ran the script locally, observed transactions were generated (tested with mina from `compatible`, `berkeley` branches) and nodes to live for more than two epochs
* Requires #14990 for the test above to pass

Checklist:

- [x] Dependency versions are unchanged
  - Notify Velocity team if dependencies must change in CI
- [ ] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [x] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [x] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [x] All tests pass (CI will check this if you didn't)
- [x] Serialized types are in stable-versioned modules
- [x] Does this close issues? None, only a step towards #14709

### CODE REVIEW

**Review by dkijania — COMMENTED** [review]
I have a general question, how different is your script from mina-local-network one? After brief analysis i cannot find anything HF related that is embedded in script and is a very different from mina-local-network approach. All above leads to conclusion that this script can be considered as a duplication ? 

I accept the answer that changing mina-local-network script was too difficult or too tedious  :)

**Comment by georgeee** [comment]
Copying here discussion between Dariusz and I that happened in Slack:

> _George Agapov:_
It's a much lighter version of local network script with great emphasis on generating a config that we need for HF testing and fully contained in a hundred of lines
I thought of using that, but then realized that it's easier for me to run mina command direct 5 times


>_Dariusz Kijania:_
I thought that this was the reason :smile: I understand that completely. However, is this only for HF testing and afterwards we wouldn't need this script? If yes are there any plans to extend mina-local-network script which your ledger generation (which i really like) ? Mina-local-network script future is also unclear. Since we have minimina and local lucy there is strong competition in this area, thus I am far from any conclusions :slightly_smiling_face:

> _George Agapov:_
I see it as part of unit test for HF flow.
HF not only Berkeley in future (I'll update script slightly, removing some special cases after Berkeley is mainnet)
As for integration, same argument as above, existing script is already a giant, I prefer to keep this precise form that could be reviewed completely in a few minutes and easily modified if necessary
>
> They have different purposes, actually
Jumbo script that we have id for launching a network
My goal was to launch a single node. Actually it's two nodes, but that is solely because bootstrap logic relies on finding some peers before anything
>
> So it's more like part of a very special unit test

**Comment by georgeee** [comment]
!ci-build-me

**Review by emberian — APPROVED** [review]
This is a beautiful script, encapsulating many years of changes ... hopefully this is the most painful HF we ever endure :)

**Comment by georgeee** [comment]
!ci-build-me

**Review by dkijania — APPROVED** [review]

**Comment by georgeee** [comment]
Recent updates:

- Introduced a new option to prepare-test-ledger.sh to allow forming of a new ledger by replacing top N delegatees in the old ledger (instead of uniformly/randomly distributing keys as was previously available)
- Introduced a script to generate three ledgers for HF testing off three consequetive mainnet ledgers (next epoch, current epoch and previous epoch)
- Updated run-hf-localnet.sh script to support compatible (previous version only attempted to do so, now it actually works)
- Added a nix build profile to build mina binary with mainnet profile

File: nix/ocaml.nix [hunk 1]
```diff
@@ -99,34 +99,7 @@ let
...
-        '';
-      });
-
-      rpc_parallel = super.rpc_parallel.overrideAttrs
-        (oa: { buildInputs = oa.buildInputs ++ [ self.ctypes ]; });
-
-      mina-dev = pkgs.stdenv.mkDerivation ({
+      minaImpl = profile: pkgs.stdenv.mkDerivation ({
         pname = "mina";
         version = "dev";
```
Comment (emberian) [inline nix/ocaml.nix:104]: should this match `profile`? 🤔

File: scripts/prepare-test-ledger.sh [hunk 2]
```diff
@@ -1,86 +1,158 @@
...
 # Number of keys in ledger that won't be re-delegated
-NUM_GHOST_KEYS=${NUM_GHOST_KEYS:-0}
-KEY_BALANCE=100000
+KEY_BALANCE=${KEY_BALANCE:-1000}
+
+# Do not touch accounts with balance below $DELEGATEE_CUTOFF
+DELEGATEE_CUTOFF=${DELEGATEE_CUTOFF:-100000}
+
+# Use (approximate) normal distribution for keys
+# Works well for latger number of keys
```
Comment (emberian) [inline scripts/prepare-test-ledger.sh:13]:
```suggestion
# Works well for larger number of keys
```

File: nix/ocaml.nix (hunk 1 above)
Comment (georgeee) [reply nix/ocaml.nix:104]: What do you mean by "match"?

P.S. probably I'll take out nix changes to a separate PR

File: scripts/prepare-test-ledger.sh [hunk 3]
```diff
@@ -1,86 +1,158 @@
...
-  wget -O $ledger_file "$ledger_url"
+  ledgers_url="https://storage.googleapis.com/storage/v1/b/mina-staking-ledgers/o?maxResults=1000&prefix=$LEDGER_PREFIX"
+  echo "$ledgers_url" >&2
+  ledger_url=$(curl "$ledgers_url" | jq -r '.items | sort_by(.size|tonumber) | last.mediaLink')
+  if [[ "$ledger_url" == null ]]; then
+    echo "Couldn't fine ledger with prefix $LEDGER_PREFIX" >&2
+    exit 2
+  fi
+  curl "$ledger_url" >"$ledger_file"
+  cmp "$ledger_file" >/dev/null <<< "Ledger not found: next staking ledger is not finalized yet" && echo "Next ledger not finalized yet" >&2 && rm "$ledger_file" && exit 2
```
Comment (emberian) [inline scripts/prepare-test-ledger.sh:82]: beware: OpenSSL ships a program called `cmp` that is entirely unrelated. it is in my path before this handy one from diffutils 😱

File: scripts/prepare-test-ledger.sh [hunk 4]
```diff
@@ -1,86 +1,158 @@
...
+
+  if [[ "$REPLACE_TOP" == "" ]]; then
+    interval=$num_keys
+    if [[ "$NORM" != "" ]]; then
+      interval=$(( (RANDOM+RANDOM+RANDOM)*num_keys/32/1024/3+1 ))
+    fi
+    # Substitute delegate of some ledger keys with the BP key
+    echo "((.[range($i; $num_accounts; $interval)] | select ((.balance|tonumber) > $DELEGATEE_CUTOFF)).delegate |= \"$key\")"
+  else
+    echo "((.[] | select(.delegate == \"${TOP_KEYS[$i]}\")).delegate |= \"$key\")"
```
Comment (emberian) [inline scripts/prepare-test-ledger.sh:120]: this is a lot cleaner than i thought it was going to be, nice.

**Review by emberian — APPROVED** [review]

File: nix/ocaml.nix (hunk 1 above)
Comment (emberian) [reply nix/ocaml.nix:104]: I meant `version = profile;` I'm not sure what the implications are of having the versions match, or where (if at all) this version is used...

Comment (georgeee) [reply nix/ocaml.nix:104]: Good point. I'll investigate it tomorrow. This one I've put just the most straightforward way I could imagine to test it working and unblock Edmund

File: scripts/prepare-test-ledger.sh (hunk 4 above)
Comment (georgeee) [reply scripts/prepare-test-ledger.sh:120]: I have a version that's even cleaner (and twice shorter in code), however for some reason it's twice slower in terms of performance (and this version takes ~15 minutes on my laptop, so it matters).

Surprisingly, replacing keys is much more expensive than randomly overwriting them based on index ranges.

File: nix/ocaml.nix (hunk 1 above)
Comment (georgeee) [reply nix/ocaml.nix:104]: I re-implemented this change in a more nix-conventional way in #15026

**Comment by georgeee** [comment]
!ci-build-me
